automata/fa/dfa.py

```diff
@@ -652,119 +569,146 @@ def complement(self, *, retain_names: bool = False, minify: bool = True) -> DFA:
             allow_partial=self.allow_partial,
         )
 
-    # visited_set, product_transitions, product_initial_state_name, final_states
-    DFADataTupleT = Tuple[Set[DFAStateT], DFATransitionsT, DFAStateT, Set[DFAStateT]]
+    @staticmethod
+    def _bfs_edges(
+        initial_state: DFAStateT,
+        expand_state_fn: ExpandStateFn,
+    ) -> Iterable[Tuple[DFAStateT, DFASymbolT, DFAStateT]]:
+        """
+        Emits the edges (src_state, label, tgt_state) visited by BFS from the
+        initial_state. Computes subsequent states using the function expand_state_fn.
+        """
+        visited_set = {initial_state}
+        queue: Deque[Tuple[DFAStateT, DFAStateT]] = deque([initial_state])
 
-    @overload
-    def _cross_product(
-        self,
-        other: DFA,
-        state_target_fn: Callable[[Tuple[DFAStateT, DFAStateT]], bool],
-        *,
-        should_construct_dfa: Literal[True],
-        retain_names: bool = False,
-    ) -> DFADataTupleT:
-        ...
+        while queue:
+            curr_state = queue.popleft()
 
-    @overload
-    def _cross_product(
-        self,
-        other: DFA,
-        state_target_fn: Callable[[Tuple[DFAStateT, DFAStateT]], bool],
-        *,
-        should_construct_dfa: Literal[False],
-        retain_names: bool = False,
-    ) -> bool:
-        ...
+            for chr, tgt_state in expand_state_fn(curr_state):
+                yield curr_state, chr, tgt_state
 
-    def _cross_product(
-        self,
-        other: DFA,
-        state_target_fn: Callable[[Tuple[DFAStateT, DFAStateT]], bool],
-        *,
-        should_construct_dfa: Literal[True, False],
-        retain_names: bool = False,
-    ) -> Union[bool, DFADataTupleT]:
+                if tgt_state not in visited_set:
+                    visited_set.add(tgt_state)
+                    queue.append(tgt_state)
+
+    @staticmethod
+    def _bfs_states(
+        initial_state: DFAStateT, expand_state_fn: ExpandStateFn
+    ) -> Iterable[DFAStateT]:
```

Comment:
I think it's technically more precise to annotate this as `Generator[DFAStateT, None, None` instead of just an iterable, since an iterable could be just a list or something.

Reply:
@EduardoGoulart1 Agreed with @eliotwrobson, because the direct return value of this function is technically a generator object.

Reply:
I switched to `Iterator` because this is actually what the interface expects. A generator is also an iterator. But if the function would return lists or iterators it would work just fine (in fact, this is what the lambdas in lines 342, 549 and 744 do).

Reply:
But not sure how to fix the current link error. Because it complains that I am importing something which I am not :question:

Reply:
@EduardoGoulart1 Seems like it might be a flake8 bug. I've committed a `# noqa` command to your branch, so CI should pass now.

Reply:
@EduardoGoulart1 I also pushed a commit with the `Generator` tweak for `_bfs_edges`, so I think we can mark this thread as resolved now. cc @eliotwrobson